PR: https://github.com/odoo/odoo/pull/206399

From: 919c5515d60b9c53faf56fa63907a1f0f5d87a64
From: Antoine Boonen
Date: 2025-06-16 13:25:50

Breaking data model changes scores: del:2 + add:0, change matches:
-    group_show_sale_receipts = fields.Boolean(string='Sale Receipt',
-    group_show_purchase_receipts = fields.Boolean(string='Purchase Receipt',

Total Changes: 48

[IMP] account: Ease receipt UI/UX

Problem
*--------
Currently, receipts are well implemented but highly hidden in the depth
of Odoo. Users are required to activate the option to see and create
receipts.

Objective
*--------
Make it easier for users to create and use receipts.

Solution
*--------
* Create a new JS widget that extends the `radio` widget to only display
a sub-range of move types on the move form:
    - From `in_invoice` to `in_receipt` and inversely;
    - From `out_invoice` to `out_receipt` and inversely.
And remove the old views as those are not necessary anymore.
* Add filters to the list view to easily retreive the receipts.
* Remove the two different security groups; receipts are now always
available.
* Add company default receipt taxes that will be used by default on
receipt (over product and account default taxes) to, once again, ease
the user experience. Those default taxes are not available from the UI
because we don't want to cumbersome the settings any more; instead, they
shall be set up in the localization company templates.

task-4677472

closes odoo/odoo#206399

Related: odoo/enterprise#83573
Related: odoo/upgrade#7573
Signed-off-by: William André (wan) <wan@odoo.com>

================================= pseudo patch: =================================

--- a/addons/account/models/__init__.py
+++ b/addons/account/models/__init__.py
@@ -38,6 +38,7 @@ from . import digest
 from . import res_users
 from . import ir_attachment
 from . import ir_actions_report
+from . import ir_http
 from . import ir_module
 from . import mail_message
 from . import mail_template

--- a/addons/account/models/account_move.py
+++ b/addons/account/models/account_move.py
@@ -5748,6 +5748,9 @@ class AccountMove(models.Model):
     def is_entry(self):
         return self.move_type == 'entry'
 
+    def is_receipt(self):
+        return self.move_type in ['out_receipt', 'in_receipt']
+
     @api.model
     def get_sale_types(self, include_receipts=False):
         return ['out_invoice', 'out_refund'] + (include_receipts and ['out_receipt'] or [])

--- a/addons/account/models/account_move_line.py
+++ b/addons/account/models/account_move_line.py
@@ -886,18 +886,24 @@ class AccountMoveLine(models.Model):
         self.ensure_one()
 
         company_domain = self.env['account.tax']._check_company_domain(self.move_id.company_id)
-        if self.move_id.is_sale_document(include_receipts=True):
-            # Out invoice.
-            filtered_taxes_id = self.product_id.taxes_id.filtered_domain(company_domain)
-            tax_ids = filtered_taxes_id or self.account_id.tax_ids.filtered(lambda tax: tax.type_tax_use == 'sale')
+        tax_ids = self.env['account.tax']
 
-        elif self.move_id.is_purchase_document(include_receipts=True):
-            # In invoice.
-            filtered_supplier_taxes_id = self.product_id.supplier_taxes_id.filtered_domain(company_domain)
-            tax_ids = filtered_supplier_taxes_id or self.account_id.tax_ids.filtered(lambda tax: tax.type_tax_use == 'purchase')
+        if self.move_id.is_receipt():
+            tax_ids = self.company_id.account_purchase_receipt_tax_id if self.move_id.move_type == 'in_receipt' else self.company_id.account_sale_receipt_tax_id
 
-        else:
-            tax_ids = False if self.env.context.get('skip_computed_taxes') or self.move_id.is_entry() else self.account_id.tax_ids
+        if not tax_ids:
+            if self.move_id.is_sale_document(include_receipts=True):
+                # Out invoice.
+                filtered_taxes_id = self.product_id.taxes_id.filtered_domain(company_domain)
+                tax_ids = filtered_taxes_id or self.account_id.tax_ids.filtered(lambda tax: tax.type_tax_use == 'sale')
+
+            elif self.move_id.is_purchase_document(include_receipts=True):
+                # In invoice.
+                filtered_supplier_taxes_id = self.product_id.supplier_taxes_id.filtered_domain(company_domain)
+                tax_ids = filtered_supplier_taxes_id or self.account_id.tax_ids.filtered(lambda tax: tax.type_tax_use == 'purchase')
+
+            else:
+                tax_ids = False if self.env.context.get('skip_computed_taxes') or self.move_id.is_entry() else self.account_id.tax_ids
 
         if self.company_id and tax_ids:
             tax_ids = tax_ids._filter_taxes_by_company(self.company_id)

--- a/addons/account/models/company.py
+++ b/addons/account/models/company.py
@@ -122,6 +122,8 @@ class ResCompany(models.Model):
     transfer_account_code_prefix = fields.Char(string='Prefix of the transfer accounts')
     account_sale_tax_id = fields.Many2one('account.tax', string="Default Sale Tax", check_company=True)
     account_purchase_tax_id = fields.Many2one('account.tax', string="Default Purchase Tax", check_company=True)
+    account_sale_receipt_tax_id = fields.Many2one('account.tax', string="Default Sale Receipt Tax", check_company=True)
+    account_purchase_receipt_tax_id = fields.Many2one('account.tax', string="Default Purchase Receipt Tax", check_company=True)
     tax_calculation_rounding_method = fields.Selection([
         ('round_per_line', 'Round per Line'),
         ('round_globally', 'Round Globally'),

--- a/None
+++ b/addons/account/models/ir_http.py
@@ -0,0 +1,11 @@
+from odoo import api, models
+
+
+class IrHttp(models.AbstractModel):
+    _inherit = 'ir.http'
+
+    @api.model
+    def lazy_session_info(self):
+        res = super().lazy_session_info()
+        res['show_sale_receipts'] = self.env['ir.config_parameter'].sudo().get_param('account.show_sale_receipts')
+        return res

--- a/addons/account/models/res_config_settings.py
+++ b/addons/account/models/res_config_settings.py
@@ -78,10 +78,7 @@ class ResConfigSettings(models.TransientModel):
         help="Intermediary account used when moving from a liquidity account to another.")
     module_account_accountant = fields.Boolean(string='Accounting')
     group_cash_rounding = fields.Boolean(string="Cash Rounding", implied_group='account.group_cash_rounding')
-    group_show_sale_receipts = fields.Boolean(string='Sale Receipt',
-        implied_group='account.group_sale_receipts')
-    group_show_purchase_receipts = fields.Boolean(string='Purchase Receipt',
-        implied_group='account.group_purchase_receipts')
+    show_sale_receipts = fields.Boolean(string='Sale Receipt', config_parameter='account.show_sale_receipts')
     module_account_budget = fields.Boolean(string='Budget Management')
     module_account_payment = fields.Boolean(string='Invoice Online Payment')
     module_account_reports = fields.Boolean("Dynamic Reports")
